mutex.go

```diff
@@ -60,21 +68,36 @@ func Acquire(spec Spec) (Releaser, error) {
 		timeout = spec.Clock.After(spec.Timeout)
 	}
 
-	for {
-		impl, err := acquire(spec.Name)
-		if err == nil {
-			return impl, nil
-		} else if err != errLocked {
-			return nil, errors.Trace(err)
-		}
-		select {
-		case <-timeout:
-			return nil, ErrTimeout
-		case <-spec.Cancel:
-			return nil, ErrCancelled
-		case <-spec.Clock.After(spec.Delay):
-			// no-op, continue and try again
-		}
+	m1, err := acquire(spec, timeout)
+	if err != nil {
+		return nil, err
+	}
+
+	// NOTE(axw) for compatibility with older versions of Juju,
+	// we must also acquire a legacy mutex. We acquire the new
+	// version first, since that provides the preferred blocking
+	// behaviour.
+	m2, err := acquireLegacy(
```

Comment:
Have we done any testing around what happens with racing new-and-old? I don't think we can deadlock, so we're probably fine. But just want to make sure we thought that part through.

Reply:
I have tested it manually. I'll try and write unit tests for that tomorrow.

Reply:
Added tests for Linux and Windows.